File: day02.py
#!/usr/bin/env python3

from get_input import get_input, line_parser

def paper(l,w,h):
    sides = (l*w, w*h, h*l)
    return 2 * sum(sides) + min(sides)

def part1(lines):
    return sum(paper(*line) for line in lines)

def ribbon(l,w,h):
    s,m,l = sorted((l,w,h))
    return 2 * (s + m) + s * m * l

def part2(lines):
    return sum(ribbon(*line) for line in lines)

def parse(line):
    return tuple(int(n) for n in line.split('x'))

if __name__ == '__main__':
    lines = line_parser(get_input(day=2, year=2015), parse=parse)
    print("Part 1: {}".format(part1(lines)))
    print("Part 2: {}".format(part2(lines)))

File: day02.rs
use common::get_input;

type Input = Vec<Present>;

struct Present {
    edges: Vec<usize>,
}

impl Present {
    fn from_string(text: &str) -> Present {
        Present { edges: text.split("x").map(|n| n.parse::<usize>().unwrap()).collect() }
    }

    fn ribbon(&self) -> usize {
        let mut edges = self.edges.clone();
        edges.sort();
        2 * (edges[0] + edges[1])  + edges[0]*edges[1]*edges[2]
    }

    fn paper(&self) -> usize {
        let edges =  &self.edges;
        let sides = [edges[0]*edges[1], edges[1]*edges[2], edges[0]*edges[2]];
        2*(sides.iter().sum::<usize>()) + sides.iter().min().unwrap()
        
    }
}

fn part1(packages: &Input) -> usize {
    packages.iter().map(|p| p.paper()).sum()
}

fn part2(packages: &Input) -> usize {
    packages.iter().map(|p| p.ribbon()).sum()
}

fn parse(text: String) -> Vec<Present> {
    text.trim().split("\n").map(|p| Present::from_string(p)).collect()
}

fn main() {
    let input = parse(get_input(02, 2015));
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}
